checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallConsistencyAnalyzer.java

```diff
@@ -0,0 +1,1703 @@
+          List<String> enclosingCmcfValues =
+              ResourceLeakVisitor.getCreatesMustCallForValues(
+                  enclosingMethodElt, mcAtf, typeFactory);
+          if (!enclosingCmcfValues.isEmpty()) {
+            for (String enclosingCmcfValue : enclosingCmcfValues) {
+              JavaExpression enclosingTarget;
+              try {
+                enclosingTarget =
+                    StringToJavaExpression.atMethodBody(
+                        enclosingCmcfValue, enclosingMethodTree, checker);
+              } catch (JavaExpressionParseException e) {
+                // Do not issue an error here, because it would be a duplicate.
+                // The error will be issued by the Transfer class of the checker,
+                // via the CreatesMustCallForElementSupplier interface.
+                enclosingTarget = null;
+              }
+
+              if (representSame(target, enclosingTarget)) {
+                // This satisifies case 3.
+                validInvocation = true;
+              }
+            }
+          }
+        }
+      }
+      if (!validInvocation) {
+        missing.add(target);
+      }
+    }
+
+    if (missing.isEmpty()) {
+      // all targets were valid
+      return;
+    }
+
+    String missingStrs = StringsPlume.join(", ", missing);
+    checker.reportError(node.getTree(), "reset.not.owning", missingStrs);
+  }
+
+  /**
+   * Checks whether the two JavaExpressions are the same. This is identical to calling equals() on
+   * one of them, with two exceptions: the second expression can be null, and "this" references are
+   * compared using their underlying type. (ThisReference#equals always returns true, which is
+   * probably a bug and isn't accurate in the case of nested classes.)
+   *
+   * @param target a JavaExpression
+   * @param enclosingTarget another, possibly null, JavaExpression
+   * @return true iff they represent the same program element
+   */
+  private boolean representSame(JavaExpression target, @Nullable JavaExpression enclosingTarget) {
+    if (enclosingTarget == null) {
+      return false;
+    }
+    if (enclosingTarget instanceof ThisReference && target instanceof ThisReference) {
+      return enclosingTarget.getType().toString().equals(target.getType().toString());
+    } else {
+      return enclosingTarget.equals(target);
+    }
+  }
+
+  /**
+   * Given a node representing a method or constructor call, checks that if the call has a non-empty
+   * {@code @MustCall} type, then its result is pseudo-assigned to some location that can take
+   * ownership of the result. Searches for the set of same resources in {@code obligations} and adds
+   * the new LocalVarWithTree to it if one exists. Otherwise creates a new set.
+   *
+   * @param obligations the currently-tracked obligations. This is always side-effected: an
+   *     obligation is either modified to include a new resource alias (the result of the invocation
```

Comment:
How an an obligation be modified?  An obligation is an ImmutableSet.  (Or is an obligation a LocalVarWithTree, in which case `obligations` isn't a set of obligations but a set of sets of obligations?)

Reply:
Technically, the old obligation is removed and a new, related one is created and added.

I don't know why we chose to make each obligation an immutable set and then remove and add a new one rather than modify them. Conceptually, modification makes sense. @msridhar do you know why?

Reply:
> Technically, the old obligation is removed and a new, related one is created and added.

Yes, that's what I see happening in the code.  I would word the Javadoc that way, or at least clarify the conceptual "obligation is modified" text with a description of what actually happens.  While reading that Javadoc, the ImmutableSet type of the obligations is visible to the reader, which is why I think a clarification is necessary.